src/asynch/wifi.rs

```diff
@@ -0,0 +1,308 @@
+
+    /// Restarts the module and blocks until ready
+    fn restart(&mut self) -> impl core::future::Future<Output = Result<(), Self::RestartError>>;
+}
+
+pub struct InnerAdapter<
+    'urc_sub,
+    A: AtatClient,
+    const TX_SIZE: usize,
+    const RX_SIZE: usize,
+    const URC_CAPACITY: usize,
+> {
+    /// ATAT client
+    pub(crate) client: A,
+
+    /// URC message subscriber
+    pub(crate) urc_subscription: UrcSubscription<'urc_sub, URCMessages<RX_SIZE>, URC_CAPACITY, 1>,
+
+    /// Network state
+    pub(crate) session: Session<RX_SIZE>,
+
+    /// Timeout for data transmission
+    pub(crate) send_timeout: Duration,
+}
+
+impl<
+        'urc_sub,
+        A: AtatClient,
+        const TX_SIZE: usize,
+        const RX_SIZE: usize,
+        const URC_CAPACITY: usize,
+    > InnerAdapter<'urc_sub, A, TX_SIZE, RX_SIZE, URC_CAPACITY>
+{
+    pub fn new(
+        client: A,
+        urc_subscription: UrcSubscription<'urc_sub, URCMessages<RX_SIZE>, URC_CAPACITY, 1>,
+        send_timeout: Duration,
+    ) -> Self {
+        Self {
+            client,
+            urc_subscription,
+            session: Session::default(),
+            send_timeout,
+        }
+    }
+
+    pub(crate) fn process_urc_messages(&mut self) {
+        while let Some(message) = self.urc_subscription.try_next_message_pure() {
+            self.session.handle_urc(message)
+        }
+    }
+
+    async fn set_station_mode(&mut self) -> Result<(), JoinError> {
+        let command = WifiModeCommand::station_mode();
+        self.send_command(command).await?;
+
+        Ok(())
+    }
+
+    async fn connect_access_point(&mut self, ssid: &str, key: &str) -> Result<(), JoinError> {
+        if ssid.len() > 32 {
+            return Err(JoinError::InvalidSSDLength);
+        }
+
+        if key.len() > 63 {
+            return Err(JoinError::InvalidPasswordLength);
+        }
+
+        let command = AccessPointConnectCommand::new(String::from_str(ssid).unwrap(), String::from_str(key).unwrap());
+        self.send_command(command).await?;
+
+        Ok(())
+    }
+
+    pub(crate) async fn send_command<Cmd: AtatCmd + CommandErrorHandler>(
+        &mut self,
+        command: Cmd,
+    ) -> Result<Cmd::Response, Cmd::Error> {
+        self.client.send(&command).await.map_err(|e| command.command_error(e))
+    }
+
+    async fn enable_multiple_connections(&mut self) -> Result<(), Error> {
+        if self.session.multi_connections_enabled {
+            return Ok(());
+        }
+
+        self.send_command(SetMultipleConnectionsCommand::multiple()).await?;
+        self.session.multi_connections_enabled = true;
+        Ok(())
+    }
+
+    fn open_socket(&mut self) -> Result<Socket, Error> {
+        if let Some(link_id) = self.session.get_next_open() {
+            self.session.sockets[link_id].state = ConnectionState::Open;
+            return Ok(Socket::new(link_id));
+        }
+
+        Err(Error::NoSocketAvailable)

```

Comment:
Same here. Logic is identical with sync part.
So maybe we could put an abstracted version in Session struct.